lib/active_model/serializer/attributes.rb

```diff
@@ -0,0 +1,109 @@
+module ActiveModel
+  class Serializer
+    module Attributes
+      class Attribute
+        delegate :call, to: :reader
+
+        attr_reader :name, :reader
+
+        def initialize(name)
+          @name = name
+          @reader = :no_reader
+        end
+
+        def self.build(name, block)
+          if block
+            AttributeBlock.new(name, block)
+          else
+            AttributeReader.new(name)
+          end
+        end
+      end
+      class AttributeReader < Attribute
+        def initialize(name)
+          super(name)
+          @reader = ->(instance) { instance.read_attribute_for_serialization(name) }
+        end
+      end
+      class AttributeBlock < Attribute
+        def initialize(name, block)
+          super(name)
+          @reader = ->(instance) { instance.instance_eval(&block) }
+        end
+      end
```

Comment:
Do you think it would make sense for you to make a PR into this branch so we can review it that way?  Sounds like we're both ok with this, except a few not-too-major things